source/common/network/udp_listener_impl.cc

```diff
@@ -112,56 +122,65 @@ void UdpListenerImpl::handleReadCallback() {
-      ASSERT(AF_INET == sin->sin_family);
-      peer_address = std::make_shared<Address::Ipv4Instance>(sin);
-
-      break;
-    }
-    case AF_INET6: {
-      const struct sockaddr_in6* sin6 = reinterpret_cast<const struct sockaddr_in6*>(&addr);
-      ASSERT(AF_INET6 == sin6->sin6_family);
-      if (IN6_IS_ADDR_V4MAPPED(&sin6->sin6_addr)) {
-#if defined(__APPLE__)
-        struct sockaddr_in sin = {
-            {}, AF_INET, sin6->sin6_port, {sin6->sin6_addr.__u6_addr.__u6_addr32[3]}, {}};
-#else
-        struct sockaddr_in sin = {AF_INET, sin6->sin6_port, {sin6->sin6_addr.s6_addr32[3]}, {}};
-#endif
-        peer_address = std::make_shared<Address::Ipv4Instance>(&sin);
-      } else {
-        peer_address = std::make_shared<Address::Ipv6Instance>(*sin6, true);
-      }
-
-      break;
-    }
-
-    default:
-      RELEASE_ASSERT(false,
-                     fmt::format("Unsupported address family: {}, local address: {}, receive size: "
-                                 "{}, address length: {}",
-                                 addr.ss_family, local_address->asString(), recv_result.result_.rc_,
-                                 addr_len));
-      break;
+    try {
+      peer_address = Address::addressFromSockAddr(
+          addr, addr_len, local_address->ip()->version() == Address::IpVersion::v6);
+    } catch (const EnvoyException&) {
+      // Intentional no-op. The assert should fail below
     }
 
     RELEASE_ASSERT((peer_address != nullptr),
                    fmt::format("Unable to get remote address for fd: {}, local address: {} ",
                                socket_.ioHandle().fd(), local_address->asString()));
 
-    RELEASE_ASSERT((local_address != nullptr),
-                   fmt::format("Unable to get local address for fd: {}", socket_.ioHandle().fd()));
+    // Unix domain sockets are not supported
+    RELEASE_ASSERT(peer_address->type() == Address::Type::Ip,
+                   fmt::format("Unsupported peer address: {} local address: {}, receive size: "
+                               "{}, address length: {}",
+                               peer_address->asString(), local_address->asString(),
+                               recv_result.result_.rc_, addr_len));
 
-    cb_.onData(UdpData{local_address, peer_address, std::move(recv_result.buffer_)});
+    UdpRecvData recvData{local_address, peer_address, std::move(recv_result.buffer_)};
+    cb_.onData(recvData);
 
   } while (true);
 }
 
-void UdpListenerImpl::handleWriteCallback() { cb_.onWriteReady(socket_); }
+void UdpListenerImpl::handleWriteCallback() {
+  ENVOY_UDP_LOG(trace, "handleWriteCallback");
+  cb_.onWriteReady(socket_);
+}
+
+Event::Dispatcher& UdpListenerImpl::dispatcher() { return dispatcher_; }
+
+const Address::InstanceConstSharedPtr& UdpListenerImpl::localAddress() const {
+  return socket_.localAddress();
+}
+
+void UdpListenerImpl::send(const UdpSendData& send_data) {
```

Comment:
I have a general question on how we are handling datagram max MTU size. Right now we don't. Should we actually expose path MTU information to the caller? Maybe a TODO in this area at the interface level?

Reply:
That's right. I will add that to the send API in the interface